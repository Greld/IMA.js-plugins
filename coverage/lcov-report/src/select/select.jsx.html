<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/select/select.jsx</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> / <a href="index.html">src/select</a> select.jsx
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">71.11% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>32/45</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">62.5% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>15/24</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">91.67% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>11/12</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>11/11</span>
      </div>
    </div>
  </div>
  <div class='status-line medium'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56</td><td class="line-coverage quiet"><span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-yes">5x</span>
<span class="cline-any cline-yes">5x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import hoistNonReactStatic from 'hoist-non-react-statics';<span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="fstat-no" title="function not covered" ><span class="branch-1 cbranch-no" title="branch not covered" ></span></span></span></span></span></span></span></span>
import AbstractPureComponent from 'ima/page/AbstractPureComponent';
import * as helpers from 'ima/page/componentHelpers';
import React from 'react';
import { createSelector } from 'reselect';
&nbsp;
export default function select(...selectors) {
	const stateSelector = createStateSelector(...selectors);
&nbsp;
	return (Component) =&gt; {
		class SelectState extends AbstractPureComponent {
			static get contextTypes() {
				return helpers.getContextTypes(this);
			}
&nbsp;
			constructor(props, context) {
				super(props, context);
			}
&nbsp;
			render() {
				let extraProps = stateSelector(this);
&nbsp;
				return &lt;Component {...extraProps} {...this.props}/&gt;;
			}
		}
&nbsp;
		hoistNonReactStatic(SelectState, Component);
&nbsp;
		return SelectState;
	};
}
&nbsp;
export function createStateSelector(...selectors) {<span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="cstat-no" title="statement not covered" ><span class="missing-if-branch" title="else path not taken" >E</span></span></span></span></span></span></span></span>
	return createSelector(
		...(selectors.map((selector) =&gt; {
			return (componentInstance) =&gt; {
			<span class="missing-if-branch" title="else path not taken" >E</span>	if ($Debug) {
					if (!componentInstance ||
						!componentInstance.context ||
						!componentInstance.context.$Utils ||
						!componentInstance.context.$Utils.$PageStateManager
					) {
						throw new Error('The $PageStateManager must be added to $Utils in component context. The $Utils are defined in app/config/bind.js.');
					}
				}
				const state = componentInstance.context.$Utils.$PageStateManager.getState();
&nbsp;
				return selector(state, componentInstance.context);
			}
		})),
		(...rest) =&gt; {
			return Object.assign({}, ...rest);
		}
	);
}
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Oct 04 2017 18:46:40 GMT+0200 (CEST)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
