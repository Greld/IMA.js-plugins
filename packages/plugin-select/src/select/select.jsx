import hoistNonReactStaticMethod from 'hoist-non-react-statics';
import AbstractPureComponent from 'ima/page/AbstractPureComponent';
import Events from 'ima/page/state/Events';
import * as helpers from 'ima/page/componentHelpers';
import React from 'react';
import { createSelector } from 'reselect';

let creatorOfStateSelector = createStateSelector;
let hoistStaticMethod = hoistNonReactStaticMethod;

export function setCreatorOfStateSelector(createStateSelector) {
  creatorOfStateSelector = createStateSelector;
}

export const hoistNonReactStatic = hoistNonReactStaticMethod;

export function setHoistStaticMethod(method) {
  hoistStaticMethod = method;
}

export default function select(...selectors) {
  return Component => {
    class SelectState extends AbstractPureComponent {
      static get contextTypes() {
        return helpers.getContextTypes(this);
      }

      constructor(props, context) {
        super(props, context);

        this.stateSelector = creatorOfStateSelector(...selectors);
        this.state = this._resolveNewState();
      }

      _resolveNewState() {
        return this.stateSelector(
          this.utils.$PageStateManager.getState(),
          this.context
        );
      }

      componentDidMount() {
        this.utils.$Dispatcher.listen(
          Events.AFTER_CHANGE_STATE,
          this.afterChangeState,
          this
        );
      }

      componentWillUnmount() {
        this.utils.$Dispatcher.unlisten(
          Events.AFTER_CHANGE_STATE,
          this.afterChangeState,
          this
        );
      }

      afterChangeState() {
        this.setState(this._resolveNewState());
      }

      render() {
        return <Component {...this.state} {...this.props} />;
      }
    }

    hoistStaticMethod(SelectState, Component);

    return SelectState;
  };
}

export function createStateSelector(...selectors) {
  const derivedState = createSelector(
    ...selectors.map(selector => {
      return (state, context) => {
        return selector(state, context);
      };
    }),
    (...rest) => {
      return Object.assign({}, ...rest);
    }
  );

  const passStateOnChange = (() => {
    let memoizedSelector = null;
    let selectorFunctions = null;
    let memoizedState = null;

    return state => {
      memoizedState = state;
      if (!selectorFunctions) {
        selectorFunctions = Object.keys(state).map(key => {
          return currentState => {
            return currentState[key] || false;
          };
        });
      }

      if (!memoizedSelector) {
        memoizedSelector = createSelector(
          ...selectorFunctions,
          () => {
            return memoizedState;
          }
        );
      }

      return memoizedSelector(state);
    };
  })();

  return createSelector(
    derivedState,
    passStateOnChange
  );
}
