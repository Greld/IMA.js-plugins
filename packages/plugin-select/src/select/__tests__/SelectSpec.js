import { shallow } from 'enzyme';
import PageStateManager from 'ima/page/state/PageStateManager';
import Dispatcher from 'ima/event/Dispatcher';
import React from 'react';
import { toMockedInstance, setGlobalMockMethod } from 'to-mock';
import select, {
  createStateSelector,
  setCreatorOfStateSelector,
  setHoistStaticMethod,
  hoistNonReactStatic
} from '../select';

setGlobalMockMethod(jest.fn);

describe('plugin-select:', () => {
  const appState = {
    media: {
      width: 90,
      height: 60
    },
    title: 'title'
  };
  const componentContext = {
    $Utils: {
      $PageStateManager: toMockedInstance(PageStateManager, {
        getState: () => {
          return appState;
        }
      }),
      $Dispatcher: toMockedInstance(Dispatcher)
    }
  };
  const selectorMethods = [
    state => {
      return {
        width: state.media.width
      };
    },
    state => {
      return {
        height: state.media.height
      };
    }
  ];

  const selectorUsingProps = (state, context, props) => ({
    width: state.media.width * props.multiplier,
    height: state.media.height * props.multiplier
  });

  beforeEach(() => {
    global.$Debug = true;
  });

  afterEach(() => {
    delete global.$Debug;
  });

  describe('createStateSelector', () => {
    it('should select extra properties from state', () => {
      let extraProps = createStateSelector(...selectorMethods)(
        appState,
        componentContext
      );

      expect(extraProps).toMatchSnapshot();
    });

    it('should return same output for same input', () => {
      let selector = createStateSelector(...selectorMethods);

      let obj1 = selector(appState, componentContext);
      let obj2 = selector(appState, componentContext);

      expect(obj1 === obj2).toBeTruthy();
    });

    it('should trow error for undefined $PageStateManager', () => {
      expect(() => {
        createStateSelector(...selectorMethods)({ context: { $Utils: {} } });
      }).toThrow();
    });
  });

  describe('select', () => {
    let wrapper = null;
    const defaultProps = {
      props: 'props',
      multiplier: 0.5
    };

    class Component extends React.PureComponent {
      static defaultProps() {
        return defaultProps;
      }

      render() {
        return <h1>text</h1>;
      }
    }

    it('should render component', () => {
      wrapper = shallow(React.createElement(Component, defaultProps), {
        context: componentContext
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('should render component with extraProps', () => {
      let EnhancedComponent = select(...selectorMethods)(Component);

      wrapper = shallow(React.createElement(EnhancedComponent, defaultProps), {
        context: componentContext
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('should render component with extraProps modifies by ownProps', () => {
      let EnhancedComponent = select(...selectorMethods, selectorUsingProps)(
        Component
      );

      wrapper = shallow(React.createElement(EnhancedComponent, defaultProps), {
        context: componentContext
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('should add listener to dispatcher after mounting to DOM', () => {
      let EnhancedComponent = select(...selectorMethods)(Component);

      wrapper = shallow(React.createElement(EnhancedComponent, defaultProps), {
        context: componentContext
      });

      wrapper.instance().componentDidMount();

      expect(componentContext.$Utils.$Dispatcher.listen).toHaveBeenCalled();
    });

    it('should remove listener to dispatcher before unmounting from DOM', () => {
      let EnhancedComponent = select(...selectorMethods)(Component);

      wrapper = shallow(React.createElement(EnhancedComponent, defaultProps), {
        context: componentContext
      });

      wrapper.instance().componentWillUnmount();

      expect(componentContext.$Utils.$Dispatcher.unlisten).toHaveBeenCalled();
    });

    it('should render component with extraProps and own createStateSelector', () => {
      setCreatorOfStateSelector((...selectors) => {
        return (state, context) => {
          return selectors.reduce((result, selector) => {
            return Object.assign({}, result, selector(state, context));
          }, {});
        };
      });
      let EnhancedComponent = select(...selectorMethods)(Component);

      wrapper = shallow(React.createElement(EnhancedComponent, defaultProps), {
        context: componentContext
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('should render component with extraProps and own static methods', () => {
      setHoistStaticMethod((TargetComponent, Original) => {
        const keys = Object.getOwnPropertyNames(Original);

        keys.forEach(key => {
          if (key === 'defaultProps') {
            const descriptor = Object.getOwnPropertyDescriptor(Original, key);
            try {
              Object.defineProperty(TargetComponent, key, descriptor);
            } catch (e) {} // eslint-disable-line no-empty
          }
        });

        return hoistNonReactStatic(TargetComponent, Original);
      });
      let EnhancedComponent = select(...selectorMethods)(Component);

      wrapper = shallow(React.createElement(EnhancedComponent, defaultProps), {
        context: componentContext
      });

      expect(typeof EnhancedComponent.defaultProps === 'function').toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
