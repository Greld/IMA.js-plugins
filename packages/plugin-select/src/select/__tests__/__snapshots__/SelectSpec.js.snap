// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin-select: createStateSelector should select extra properties from state 1`] = `
Object {
  "height": 60,
  "width": 90,
}
`;

exports[`plugin-select: select should forward ref 1`] = `
<SelectState
  _forwardedRef={
    Object {
      "current": null,
    }
  }
  multiplier={0.5}
  props="props"
/>
`;

exports[`plugin-select: select should remove _forwardedRef prop (replaces it with ref) 1`] = `
<Component
  height={60}
  multiplier={0.5}
  props="props"
  width={90}
/>
`;

exports[`plugin-select: select should render component 1`] = `
<h1>
  text
</h1>
`;

exports[`plugin-select: select should render component with extraProps 1`] = `
<Component
  height={60}
  multiplier={0.5}
  props="props"
  width={90}
/>
`;

exports[`plugin-select: select should render component with extraProps and own createStateSelector 1`] = `
<Component
  height={60}
  multiplier={0.5}
  props="props"
  width={90}
/>
`;

exports[`plugin-select: select should render component with extraProps and own static methods 1`] = `
<Component
  height={60}
  multiplier={0.5}
  props="props"
  width={90}
/>
`;

exports[`plugin-select: select should render component with extraProps modifies by ownProps 1`] = `
<Component
  height={30}
  multiplier={0.5}
  props="props"
  width={45}
/>
`;
